docs/architecture.md

```diff
@@ -0,0 +1,465 @@
+   implementing serialization logic, or calling a database.
+ - Adapters should either reside in the top-level crate ("fuel-core") or in their own crates. This prevents the domains
+   from being polluted by dependencies with large compile times or unstable interfaces. This allows for libraries like 
+   networking, third-party SDKs, or database drivers to be swapped out without breaking our core domain specific logic.
+
+## Domains:
+
+```mermaid
+flowchart TB
+    subgraph Fuel Core
+        direction TB
+        subgraph POAS[PoA Service]
+            POAD{{Domain}}
+            SYNCPOA[Synchronizer Adapter]
+        end
+        subgraph BPS[Block Producer Service]
+            BPD{{Domain}}
+            POABP[PoA Adapter]
+            APIBP[API Adapter]
+        end
+        subgraph BIS[Block Importer Service]
+            BID{{Domain}}
+            SYNCBI["Sync Adapter"]
+            POABI["PoA Adapter"]
+        end
+        subgraph SS[Synchronizer Service]
+            SSD{{Domain}}
+        end
+        subgraph P2PS[P2P Service]
+            P2PD{{Domain}}
+            POAP2P[PoA Adapter]
+            SYNCP2P[Sync Adapter]
+            TXP2P[Transaction Pool Adapter]
+        end
+        subgraph TXS[Transaction Pool Service]
+            TXD{{Domain}}
+            BPTXS[Block Producer Adapter]
+            BITXS[Block Importer Adapter]
+        end
+        subgraph RS[Relayer Service]
+            RSD{{Domain}}
+            POAR[POA Adapter]
+            BPR[Block Producer Adapter]
+            BIR[Block Importer Adapter]
+        end
+        subgraph ES[Executor Service]
+            ESD{{Domain}}
+            BPEXE[Block Producer Adapter]
+            BIEXE[Block Importer Adapter]
+        end
+        subgraph APIS[API Service]
+            APID{{Domain}}
+        end
+        BID & BPD --> ES & RS & TXS
+        SSD --> POAS & P2PS & BIS
+        TXD --> P2PS
+        POAD --> P2PS & RS & BPS & BIS
+        APID --> BPS
+        
+        %% Red color - Block Importer
+        linkStyle 0,1,2 fill:none,stroke-width:4px,stroke:red
+        %% Pink color - Block Producer
+        linkStyle 3,4,5 fill:none,stroke-width:4px,stroke:pink
+        %% Green color - Synchronizer
+        linkStyle 6,7,8 fill:none,stroke-width:4px,stroke:green
+        %% Blue color - Transaction pool
+        linkStyle 9 fill:none,stroke-width:4px,stroke:blue
+        %% Orange color - PoA
+        linkStyle 10,11,12,13 fill:none,stroke-width:4px,stroke:orange
+        %% Purple color - API
+        linkStyle 14 fill:none,stroke-width:4px,stroke:purple
+    end
+```
+
+### (PoA) Consensus Module
+
+#### Ports: fuel_poa_consensus::ports
+```rust
+trait BlockProducer {
+    // used by the PoA node to produce a new block and immediately save the resultant block state
+    async fn produce_and_store_block(height: BlockHeight) -> Result<FuelBlock>;
```

Comment:
yeah I was thinking that would be more efficient from the producer perspective to avoid handling state_diffs or recomputing the block again in validation mode. That's why I was considering making the PoA module do the block broadcast instead of block importer.

Reply:
Maybe these actions could be consolidated from POA to block importer by having a separate block importer method that just commits to the SealedBlockHeaders without executing the block.

Reply:
We will face problems with state diff later when we will implement PoS. Because The block should not be committed while there are not enough votes.

So for example `async fn produce(height: BlockHeight) -> Result<(FuelBlock, StateDiff)>;` may be helpful for CM. CM calls it, receives the block, and after voting gathering, calls `commit(block, Some(state_diff))`(or a separate method).

So in the final design, we will use something like this. For now, we can do double-work and optimize it later.

Reply:
> My original goal was to allow the system to work optimally without being adversely affected by uncertainty related to components we haven't built out or designed yet. Personally, I'd prefer some kind of approach that is explicit about what is going on, rather than relying on some magic executor cache to fix things in the future for block production.

I mentioned PoS only to show that we will have this problem in the future=) My main idea is not to have several places where we can commit the block because, in this case, we need to broadcast from both places. `BlockProducer` can use `BlockCommiter` inside, but I don't see how we can do that without caching/metadata.

> Having the block producer return a StateDiff could help, but if we had a way to manage database transactions across services we could bundle other things like consensus metadata as part of one atomic commit


I forgot about the pattern with metadata=D Yea, we can use it. But then we need to have different metadata like `TrustedMetadata` and `Untrusted`, because we previously discussed that other peers could share metadata about execution to speed up the sync process of other nodes. All metadata received from the network is untrusted. All metadata received from the `BlockProducer` is trusted